src/Illuminate/Database/Eloquent/Relations/BelongsToThrough.php

```diff
@@ -0,0 +1,191 @@
+    }
+
+    /**
+     * Set the constraints for an eager load of the relation.
+     *
+     * @param  array  $models
+     * @return void
+     */
+    public function addEagerConstraints(array $models)
+    {
+        $this->getQuery()->addSelect([
+            $this->getParent()->getQualifiedKeyName().' as '.self::RELATED_THROUGH_KEY,
+        ]);
+        $this->getQuery()->whereIn($this->getParent()->getQualifiedKeyName(), $this->getKeys($models, $this->localKey));
+    }
+
+    /**
+     * Initialize the relation on a set of models.
+     *
+     * @param  \Illuminate\Database\Eloquent\Model[]  $models
+     * @param  string  $relation
+     * @return array
+     */
+    public function initRelation(array $models, $relation)
+    {
+        foreach ($models as $model) {
+            $model->setRelation($relation, $this->getRelated());
+        }
+
+        return $models;
+    }
+
+    /**
+     * Match the eagerly loaded results to their parents.
+     *
+     * @param  array  $models
+     * @param  \Illuminate\Database\Eloquent\Collection  $results
+     * @param  string  $relation
+     * @return array
+     */
+    public function match(array $models, Collection $results, $relation)
+    {
+        $dictionary = $this->buildDictionary($results);
+
+        foreach ($models as $model) {
+            $key = $model->{$this->localKey};
+
+            if (isset($dictionary[$key])) {
+                $model->setRelation($relation, $dictionary[$key]);
+            }
+        }
+
+        return $models;
+    }
+
+    /**
+     * Get the results of the relationship.
+     *
+     * @return \Illuminate\Database\Eloquent\Model
+     */
+    public function getResults()
+    {
+        return $this->getQuery()->first();
+    }
+
+    /**
+     * Set the required joins on the relation query.
+     *
+     * @return void
+     */
+    protected function setJoins()
+    {
+        $one = $this->getRelated()->getQualifiedKeyName();
+        $prev = $this->getRelated()->getForeignKey();
+        foreach ($this->models as $model) {
+            $other = $model->getTable().'.'.$prev;
+            $this->getQuery()->leftJoin($model->getTable(), $one, '=', $other);
+
+            $prev = $model->getForeignKey();
+            $one = $model->getQualifiedKeyName();
+        }
+    }
+
+    /**
+     * Set the soft deleting constraints on the relation query.
+     *
+     * @return void
+     */
+    protected function setSoftDeletes()
+    {
+        foreach ($this->models as $model) {
+            if ($this->hasSoftDeletes($model)) {
+                $this->getQuery()->whereNull($model->getQualifiedDeletedAtColumn());
+            }
+        }
+    }
+
+    /**
+     * Determine whether the model uses Soft Deletes.
+     *
+     * @param  \Illuminate\Database\Eloquent\Model  $model
+     * @return bool
+     */
+    public function hasSoftDeletes(Model $model)
+    {
+        return in_array(
+            'Illuminate\Database\Eloquent\SoftDeletes',
```

Comment:
maybe better to import this, then write `return in_array(SoftDeletes::class, class_uses_recursive(get_class($model)));`